src/metabase/actions.clj

```diff
@@ -145,39 +146,57 @@
 
 ;;;; Action definitions.
 
-;;; Common base spec for all Actions. All Actions at least require
+;;; Common base spec for *all* Actions. All Actions at least require
 ;;;
-;;;    {:database <id>, :query {:source-table <id>}}
+;;;    {:database <id>}
+;;;
+;;; Anything else required depends on the action type.
 
 (s/def :actions.args/id
   (s/and integer? pos?))
 
-(s/def :actions.args.crud.common/database
+(s/def :actions.args.common/database
   :actions.args/id)
 
-(s/def :actions.args.crud.common.query/source-table
+(s/def :actions.args/common
+  (s/keys :req-un [:actions.args.common/database]))
+
+;;; Common base spec for all CRUD row Actions. All CRUD row Actions at least require
+;;;
+;;;    {:database <id>, :query {:source-table <id>}}
+
+(s/def :actions.args.crud.row.common.query/source-table
   :actions.args/id)
 
-(s/def :actions.args.crud.common/query
-  (s/keys :req-un [:actions.args.crud.common.query/source-table]))
+(s/def :actions.args.crud.row.common/query
+  (s/keys :req-un [:actions.args.crud.row.common.query/source-table]))
 
-(s/def :actions.args.crud/common
-  (s/keys :req-un [:actions.args.crud.common/database
-                   :actions.args.crud.common/query]))
+(s/def :actions.args.crud.row/common
+  (s/merge
+   :actions.args/common
+   (s/keys :req-un [:actions.args.crud.row.common/query])))
 
 ;;; the various `:row/*` Actions all treat their args map as an MBQL query.
 
-(defn- normalize-as-mbql-query [query]
-  (let [query (assoc (mbql.normalize/normalize query)
-                     :type :query)]
-    (try
-      (schema/validate mbql.s/Query query)
-      (catch Exception e
-        (throw (ex-info
-                (ex-message e)
-                {:exception-data (ex-data e)
-                 :status-code    400}))))
-    query))
+(defn- normalize-as-mbql-query
+  "Normalize `query` as an MBQL query. Optional arg `:exclude` is a set of *normalized* keys to exclude from recursive
+  normalization, e.g. `:create-row` for the `:row/create` Action (we don't want to normalize the row input since
+  preserving case and `snake_keys` in the request body is important)."
+  ([query]
+   (let [query (mbql.normalize/normalize (assoc query :type :query))]
+     (try
+       (schema/validate mbql.s/Query query)
+       (catch Exception e
+         (throw (ex-info
+                 (ex-message e)
+                 {:exception-data (ex-data e)
+                  :status-code    400}))))
+     query))
+
+  ([query & {:keys [exclude]}]
+   (let [query (update-keys query mbql.u/normalize-token)]
+     (merge (select-keys query exclude)
+            (normalize-as-mbql-query (apply dissoc query exclude))))))
```

Comment:
Fix for #23930 -- we can now exclude keys from normalization e.g. `:create-row` for the `:row/create` action.